### PR TITLE
Add support for eager loading Active Storage variants

### PR DESCRIPTION
**Pull request by ghiculescu**
Currently Active Storage [variant tracking](https://github.com/rails/rails/pull/37901) runs a query for each attachment to check for a variant. Regular Rails n+1 prevention techniques (`includes`) cannot prevent this.

This PR adds `with_all_variant_records`, as well as allowing `includes` to work as expected on Active Storage attachments.

```ruby
user.vlogs.with_all_variant_records.map do |vlog|
  vlog.representation(resize: "100x100").processed
end
```

It also updates the builtin has_many scope to also load variant records. So code like this which currently has n+1s will not have them after this PR:

```ruby
User.where(id: user.id).with_attached_vlogs.map do |user|
  user.vlogs.map do |vlog|
    vlog.representation(resize: "100x100").processed
  end
end
```

Fixes several of the comments in https://github.com/rails/rails/pull/37901.

In implementing this, I borrowed heavily from the style of https://github.com/rails/rails/pull/39397.

### CODE REVIEW

**Comment by vizcay** [comment]
Looks interesting @ghiculescu!

This will allow to do `Product.where(id: [...]).with_attached_image.with_all_variant_records` ? (And will query once each table: products, attachments, blobs and variant_records?)

**Comment by ghiculescu** [comment]
@vizcay good idea. Just updated accordingly. `Product.where(id: [...]).with_attached_image` is all you need, see [test](https://github.com/rails/rails/pull/40842/files#diff-e2ef3d19f6324cb172e3a3b9175018f704ea4571ac334b7006453d69e717879dR129).

**Comment by rickyPanzer** [comment]
would this work if the storage is aws s3?

**Comment by ghiculescu** [comment]
@rickyPanzer this should work with any storage backend. It doesn't actually interact with it the storage service at all. If you're using [tracked variants](https://github.com/rails/rails/pull/37901), that will allow you to store in your DB a pointer to a variant in the storage backend - this PR enables you to eagerly load those pointers. But both this and variant tracking itself should be backend agnostic.

**Comment by rickyPanzer** [comment]
Just coded against this PR last night, works great! I hope it gets merged in

**Comment by ghiculescu** [comment]
@georgeclaghorn can I please ask for a review here?

**Comment by VinyLimaZ** [comment]
Some news about it? I think that's a great feature and could help a lot for pages with lots of images on AS

**Comment by ghiculescu** [comment]
@VinyLimaZ @rickyPanzer @vizcay can you share what use cases (/sample code) you have in mind for this feature? It would help me ensure that the current implementation will do what you need.

**Comment by brunoprietog** [comment]
Is there any news about this?  I was trying to do something like this:

```
@addresses.includes(user: [user_information: [avatar_attachment: [:blob, :variant_record]]])
```

I want to include the thumbnail variant so I don't have to request one by one in the serializer

Thanks!

**Comment by barefootford** [comment]
Would love to keep this PR alive. @ghiculescu I've got a page that shows a few hundred thumbnails (50px variant of original) and there doesn't seem to be a way to avoid the N+1 unless I store the thumbnails as a separate attachment. Hope this can get in.

**Comment by ghiculescu** [comment]
I'd love to keep it alive too! I think it's ready to go, I'm just not sure how to get it merged in 🤷‍♂️

**Comment by tosbourn** [comment]
> I'd love to keep it alive too! I think it's ready to go and the build is green, I'm just not sure how to get it merged in 🤷‍♂️

@ghiculescu looks like there is [one failing test](https://buildkite.com/rails/rails/builds/76096#c8500e38-3920-4d65-b476-ec3d7ea761a8) on the current build, looks like an unrelated test may need to be updated?

**Comment by ghiculescu** [comment]
@tosbourn looks like it was unrelated. I've rebased against master which should fix it 🤞

**Comment by wbharding** [comment]
[I posted a blog this morning for those who want to monkeypatch an existing Rails install to gain this functionality](https://bill.harding.blog/2021/04/15/applying-a-text-watermark-with-rails-activestorage-6-1/). See the "Extra credit" section

**Comment by ghiculescu** [comment]
@wbharding you might like the `with_all_variant_records` scope.

Any chance you can share the context around your calling code?

I think the Active Storage Way (tm) would be to do `<%= image_tag feature.screenshot.variant(gravity: "Southeast", pointsize: "30", fill: "#999", draw: "text 0,0 'NoteApps.info'") %>` in your view. Once the page is rendered, the browser will make a subsequent request to [`ActiveStorage::Representations::RedirectController`](https://github.com/rails/rails/blob/main/activestorage/app/controllers/active_storage/representations/redirect_controller.rb) which is where `.processed.url` is called.

I suspect one of the reasons this PR is being held up is because you only get N+1s if you do `feature.screenshot.variant(gravity: "Southeast", pointsize: "30", fill: "#999", draw: "text 0,0 'NoteApps.info'").processed.url` in your view/controller/etc. So I think it would be good to collect use cases for where this is necessary, to help the Rails team see the value of this PR.

**Comment by wbharding** [comment]
Hey @ghiculescu sure thing! Really appreciate the work that you've been putting into figuring out the solution to this and getting it accepted.

The use case I describe in the blog post is for my site [noteapps.info](https://www.noteapps.info), where there are a few reasons that we are using `processed.url`:

We are running a site on Heroku (high variance in response time, high variance in transit between database and query) that is trying to render up to hundreds of images per page load, [like this page](https://www.noteapps.info/apps/compare?note_app=amplenote%2Bevernote%2Bnotion%2Broam&selected_group=tasks#compare-preface), where any checkbox icon you hover on pops a screenshot. Since we our screenshot URLs use a small watermark, and are rendered within a presenter (no access to typical view methods), making a call with the deep include syntax and your improvement to `record` is the only way for us to avoid potentially tens of separate requests being made to our underpowered Heroku app server. 

Another example is the [app-specific page](https://www.noteapps.info/apps/amplenote), where we need to look up about 150 images for a particular note taking app. If it weren't for `processed.url`, then scrolling down this page would deluge our app server in requests that became database connections and spiked our request volume. Being able to grab all of the variant blobs in the controller means that even with 150 images that contain variants (the text watermark), the page still has reasonable ActiveRecord performance characteristics.

**Comment by RSO** [comment]
I'm in a similar boat as @wbharding. The `ActiveStorage::Representations::RedirectController` sounds very suited to showing images that are ok to be loaded a little later than the rest of the page. However, I'd like to render an overview of a collection and use variations to allow clients to either show the `1x` or `2x` (retina) version of the Icon using the `srcset` attribute on the `img` tag. 

If I don't use `.processed.url` in my view I can see a noticeable difference in the loading experience because each image has to go through the `RedirectController` before redirecting to the `S3` link. But maybe this also means that there's an opportunity to add better caching to the `RedirectController`? I'm not sure whether that would pose any issues 🤔 .

**Comment by ghiculescu** [comment]
> But maybe this also means that there's an opportunity to add better caching to the `RedirectController`?

Specifically on this point, have you checked out `ActiveStorage.service_urls_expire_in`?

**Comment by collimarco** [comment]
@wbharding We use a similar strategy, because we have pages (restaurant menus) with hundreds of images and we need to avoid a request to the app server for each image (we use Cloudflare and S3 directly using `variant.url`).

Time ago I posted a description of our solution here:
https://stackoverflow.com/questions/59013049/rails-activestorage-how-to-avoid-one-redirect-for-each-image/59107484#59107484

@ghiculescu This pull request should save us from N+1. 

Currently we use `image.variant().key` together with `with_attached_image`... but we are getting N+1. 

Your pull request should solve our problem. I can't wait to see this merged.

**Comment by collimarco** [comment]
I also found this other person that is using a similar strategy (and would thus benefit from this pull request):

https://dev.to/lipanski/serving-activestorage-uploads-through-a-cdn-with-rails-direct-routes-3nc2

So it seems a common use case. Would be appreciated if Rails core team could consider this pull request.

**Comment by collimarco** [comment]
@ghiculescu Is there any workaround while waiting for the merge?

I was trying this code:

```
category.items.includes(image_attachment: { blob: :variant_records }).each do |item|
  item.image.variant({ ... }).key
end
```

The above code triggers 3 queries (attachment, blob, variant) for each item... so the `includes` does not work. Do you know what can I do? Have you found a workaround?

**Comment by ghiculescu** [comment]
If you are running Rails 6.1 or edge Rails, you could point your gemfile to this branch instead. Otherwise try monkey patching the “record” method with the changes from this PR.

**Comment by Lukasz93P** [comment]
Hi, why this branch is still not merged?

**Comment by zzak** [comment]
@Lukasz93P Quoting ghiculescu here in this comment: https://github.com/rails/rails/pull/40842#issuecomment-820660716

> I suspect one of the reasons this PR is being held up is because you only get N+1s if you do feature.screenshot.variant(gravity: "Southeast", pointsize: "30", fill: "#999", draw: "text 0,0 'NoteApps.info'").processed.url in your view/controller/etc. So I think it would be good to collect use cases for where this is necessary, to help the Rails team see the value of this PR.

If you could give this branch a try and let us know any issues you run into that would be a huge help moving this forward :pray:

**Comment by collimarco** [comment]
@zzak Take a page like this https://buonmenu.com/example with hundreds of images... That's why we preprocess. We can't have a single request for each image to our Rails server (our current strategy: 1 request; Rails default strategy: 100x requests). If you need more information about technical details you can ask me, but please merge this pull request.

**Review by zzak — APPROVED** [review]
Actually looking more closely at this, I think this patch is good. 🙏

**Comment by swrobel** [comment]
@ghiculescu it looks like this can probably be merged - there's just a conflict w/ the CHANGELOG that needs to be fixed 🙏🏼

**Comment by zzak** [comment]
@swrobel I don't think I get a final say in merging this (since I'm not technically a committer), so we may have to wait for an "official" review. 🙏

**Comment by ghiculescu** [comment]
Thanks for the nudge. Merge conflict fixed, commits squashed.

**Review by mihap — APPROVED** [review]

File: activestorage/test/test_helper.rb
Comment (pixeltrix) [inline activestorage/test/test_helper.rb:78]: This could probably do with extracting along with the copies in Action Text and Action View into a module - not for this PR though.

File: activestorage/CHANGELOG.md
Comment (pixeltrix) [inline activestorage/CHANGELOG.md:4]: It appears that these scopes aren't documented anywhere - they could really do with appearing in both the API and Guides documentation. Asking you to write the documentation for other PRs is a bit much so perhaps someone can create a documentation PR for these and other scopes like `with_all_rich_text`.

**Review by pixeltrix — APPROVED** [review]

**Comment by pixeltrix** [comment]
@ghiculescu thanks 👍🏻

**Comment by collimarco** [comment]
Thanks, I'm really happy to see this merged 🎉

What Rails version will include this? Patch, minor or major release version?

**Comment by pixeltrix** [comment]
Since this changes the behaviour subtly of some existing methods it should be 7.0 but a more limited backport that just mirrors what's currently available with Action Text is possible.

File: activestorage/CHANGELOG.md
Comment (ghiculescu) [reply activestorage/CHANGELOG.md:4]: Good call. I’ll make a follow up. Thanks for the review!

**Comment by collimarco** [comment]
@pixeltrix Is there any chance to have a backport? Rails 7 seems very far and this would be a life-safer for us.

**Comment by ghiculescu** [comment]
@collimarco I’d suggest running Rails against the main branch, it’s much less scary than you’d think and means you’ll basically get all the stuff that’s gonna be in Rails 7.

**Comment by collimarco** [comment]
**Ouch**... I'm testing this PR in our application and it does not fix the N+1 issue for us :( 

This is the code:

```
category.items.with_attached_image.each do |item|
  puts "https://cdn.example.com/" + item.image.variant({ resize: "100x100" }).key
end
```

- If inside the loop we call `variant.processed` we have **only 1 query**
- If inside the loop we call `variant.key` (or `variant.processed.key`) we have **N+1 queries**

This is the log output for each iteration if you call `variant.key`:

```
ActiveStorage::Attachment Load (0.3ms)  SELECT "active_storage_attachments".* FROM "active_storage_attachments" WHERE "active_storage_attachments"."record_id" = $1 AND "active_storage_attachments"."record_type" = $2 AND "active_storage_attachments"."name" = $3 LIMIT $4  [["record_id", 10], ["record_type", "ActiveStorage::VariantRecord"], ["name", "image"], ["LIMIT", 1]]
ActiveStorage::Blob Load (0.3ms)  SELECT "active_storage_blobs".* FROM "active_storage_blobs" WHERE "active_storage_blobs"."id" = $1 LIMIT $2  [["id", 100], ["LIMIT", 1]]
```

Basically `ActiveStorage::Attachment` and `ActiveStorage::Blob` are loaded again, even though they are already available.

**Comment by ghiculescu** [comment]
https://github.com/rails/rails/pull/42981 fixes that ^

**Comment by paul-bonnel-fr** [comment]
> **Ouch**... I'm testing this PR in our application and it does not fix the N+1 issue for us :(
> 
> This is the code:
> 
> ```
> category.items.with_attached_image.each do |item|
>   puts "https://cdn.example.com/" + item.image.variant({ resize: "100x100" }).key
> end
> ```
> 
> * If inside the loop we call `variant.processed` we have **only 1 query**
> * If inside the loop we call `variant.key` (or `variant.processed.key`) we have **N+1 queries**
> 
> This is the log output for each iteration if you call `variant.key`:
> 
> ```
> ActiveStorage::Attachment Load (0.3ms)  SELECT "active_storage_attachments".* FROM "active_storage_attachments" WHERE "active_storage_attachments"."record_id" = $1 AND "active_storage_attachments"."record_type" = $2 AND "active_storage_attachments"."name" = $3 LIMIT $4  [["record_id", 10], ["record_type", "ActiveStorage::VariantRecord"], ["name", "image"], ["LIMIT", 1]]
> ActiveStorage::Blob Load (0.3ms)  SELECT "active_storage_blobs".* FROM "active_storage_blobs" WHERE "active_storage_blobs"."id" = $1 LIMIT $2  [["id", 100], ["LIMIT", 1]]
> ```
> 
> Basically `ActiveStorage::Attachment` and `ActiveStorage::Blob` are loaded again, even though they are already available.

I observe se exact same behavior on my end. This code does not seems to fix this N+1 issue. I'm running Rails '7.0.4' on my project. Is there anything I can do to help testing it ? 

```

**Pull request merged**
